spring-cloud-aws-sns/src/main/java/io/awspring/cloud/sns/core/ListTopicArnResolverCached.java

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright 2013-2023 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.awspring.cloud.sns.core;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.stream.Collectors;
+import org.springframework.util.Assert;
+import software.amazon.awssdk.arns.Arn;
+import software.amazon.awssdk.services.sns.SnsClient;
+import software.amazon.awssdk.services.sns.model.Topic;
+
+/**
+ * @author Matej Nedić
+ * @since 3.0.0
+ */
+public class ListTopicArnResolverCached implements TopicArnResolver {
+
+	private final SnsClient snsClient;
+
+	private final ConcurrentHashMap<String, Arn> cache = new ConcurrentHashMap<>();
+
+	public ListTopicArnResolverCached(SnsClient snsClient) {
+		this.snsClient = snsClient;
+		resolveByCallingPaginator();
```

Comment:
Should we initialize it at initialization?
Can this map become 2 large? Also, if it is 2 large should the application have permission to see other topics that are not in the scope of the application?
cc @maciejwalkowiak

Reply:
Well, I'll drop my 2c, hope it's ok :)

> Should we initialize it at initialization?

IMHO we should not, as this can impact the application startup time. I'd rather keep it lazy and populate the cache on the first call.

> Can this map become 2 large? 

I think only if users have too many topics, like thousands. Doesn't seem like a problem IMO.

> Also, if it is 2 large should the application have permission to see other topics that are not in the scope of the application?

I think this is more of an IAM problem than a framework problem, so I wouldn't worry.

Reply:
Even if we decide to do it on the initialization, it should not be done in constructor but rather class should implement`InitializingBean` and do it in `afterPropertiesSet()`.